### PR TITLE
Kubernetes/openshift container

### PR DESCRIPTION
**Pull request by petervo**
Depends on #3466 and #3355.

Do we want to build containers like this from RPMS, or do we want to for a tar inside the container from the tarball.

Still need to get testing to work with this.

### CODE REVIEW

File: images/kubernetes/src/cockpit-kube-auth/helpers/client.go [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+	// Treat connection errors as internal errors and invalid
+	// responses as auth errors
+	if e != nil {
+		return errors.New(fmt.Sprintf("Couldn't connect to api: %s", e))
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != 401 && resp.StatusCode != 403 {
+		return errors.New("API is open, refusing to connect")
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-auth/helpers/client.go]: "API is unauthenticated"

File: images/kubernetes/src/cockpit-kube-auth/helpers/client.go [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+		if _, ok := err.(*AuthError); ok {
+			return err
+		} else {
+			return errors.New(fmt.Sprintf("Couldn't connect to the user api endpoint: %v", err))
+		}
+	}
+
+	defer resp.Body.Close()
+
+    // If we got a 404 and this isn't token auth
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-auth/helpers/client.go]: Indentation?

File: images/kubernetes/src/cockpit-kube-auth/helpers/client.go [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+	cluster["insecure-skip-tls-verify"] = self.insecure
+	clusters := namedArrayObject("cluster", "container-cluster", cluster)
+
+	context := make(map[string]interface{})
+	context["cluster"] = "container-cluster"
+	context["user"] = user_data["name"]
+	contexts := namedArrayObject("context", "container-context", context)
+
+	users := make([]map[string]interface{}, 1)
+    users[0] = user_data
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-auth/helpers/client.go]: Indentation

File: images/kubernetes/src/cockpit-kube-auth/helpers/client.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	user_data := creds.GetApiUserMap()
+	cluster := make(map[string]interface{})
+	cluster["server"] = self.host
+	if self.caData != "" {
+    	cluster["certificate-authority-data"] = self.caData
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-auth/helpers/client.go]: Indentation

File: images/kubernetes/src/cockpit-kube-auth/helpers/client.go [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+	ac.insecure = false
+	ac.host = os.Getenv("KUBERNETES_SERVICE_HOST")
+	if ac.host != "" {
+		// assume we are always on https
+		ac.host = fmt.Sprintf("https://%s", ac.host)
+		port := os.Getenv("KUBERNETES_SERVICE_PORT")
+		if port != "" {
+			ac.host = fmt.Sprintf("%s:%s", ac.host, port)
+		}
+    }
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-auth/helpers/client.go]: Indentation.

File: images/kubernetes/src/cockpit-kube-auth/helpers/creds.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+		cred.authHeader = fmt.Sprintf("Bearer %s", cred.bearerToken)
+	} else {
+		return nil, errors.New(fmt.Sprintf("Unsuported authentication type %s", authType))
+	}
+
+	return cred, nil
+}
+
+func NewCredentialsForSystem() (*Credentials, error) {
+    token, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/token")
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-auth/helpers/creds.go]: Indentation

File: images/kubernetes/src/cockpit-kube-auth/main.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+		}
+	}
+
+	sendAuthResponse(authFd, response)
+
+	if loginErr == nil {
+	    if os.Getenv("XDG_RUNTIME_DIR") == "" {
+	        os.Setenv("XDG_RUNTIME_DIR", "/tmp")
+	    }
+		syscall.Exec("/usr/libexec/cockpit-stub", nil, os.Environ())
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-auth/main.go]: Indentation

File: pkg/kubernetes/client.js [hunk 8]
```diff
@@ -580,20 +588,27 @@ define([
...
+                    var scheme = config.parse_scheme(data);
+                    debug("kube config scheme:", scheme);
+                    schemes.unshift(scheme);
+                })
+                .always(step);
+        }
+
+        schemes.unshift(get_kubectl_config);
+        if (login_data)
+            schemes.unshift(config.parse_scheme(login_data));
```
Comment (stefwalter) [inline pkg/kubernetes/client.js]: Lets do this the other way around. Always prefer the login data.

**Comment by stefwalter** [comment]
Could you run gofmt on the files to get them in the right format?

**Comment by stefwalter** [comment]
I realized this has a lot of deps elsewhehre, so will continue reviews once those are merged and thish is rebased.

File: pkg/kubernetes/client.js (hunk 8 above)
Comment (petervo) [inline pkg/kubernetes/client.js]: I think that's what this does, unshift adds to the beginning of the array

Comment (stefwalter) [inline pkg/kubernetes/client.js]: Ah, sorry, it's late.

**Comment by petervo** [comment]
On 01/14/2016 12:10 PM, Stef Walter wrote:
> Could you run gofmt on the files to get them in the right format?
>

I did, not sure why it messed up indentation. Or were you talking about 
something else.

**Comment by petervo** [comment]
Rebased

File: images/kubernetes/src/cockpit-kube-launch/main.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+		)
+	}
+
+	if isOpenShift {
+		title = "Cockpit Openshift"
+	}
+
+	writeFile("/etc/cockpit/cockpit.conf", strings.Join(lines, "\n"))
+	writeFile("/etc/os-release", fmt.Sprintf("NAME=%s\nID=%s\nPRETTY_NAME=%s", title, id, title))
+	writeFile("/container/simple-shell/override.json", fmt.Sprintf("{ \"title\": \"%s\" }", title))
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-launch/main.go]: Can we do the above two from the Dockerfile? It seems less hard coded that way.

File: images/kubernetes/src/cockpit-kube-launch/main.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	filedirectory := filepath.Dir(filename)
+	location, e := filepath.Abs(filedirectory)
+	if e != nil {
+		log.Fatalf("Could't find location %v", e)
+	}
+
+	path := path.Join(location, "cockpit-kube-auth")
+
+	title := "Cockpit Kubernetes"
+	id := "cockpit-kubernetes"
```
Comment (stefwalter) [inline images/kubernetes/src/cockpit-kube-launch/main.go]: Can we build alternate os-release and override.json with the titles (for Openshift, Kubernetes, AEP, etc...) and move them into place?

**Comment by jscotka** [comment]
Hi,
it seems that your changes break selenium tests:

 /build/cockpit/test/avocado/phantom-driver: No such file or directory Wrote TestAvocadoSeleniumSuite-testAvocadoSeleniumFirefox-FAIL.png

File: images/kubernetes/container-files/kubernetes-brand [hunk 11]
```diff
@@ -0,0 +1,3 @@
+NAME=Cockpit Openshift
+ID=cockpit-kubernetes
+PRETTY_NAME=Cockpit Openshift
```
Comment (stefwalter) [inline images/kubernetes/container-files/kubernetes-brand]: This is backwards.

File: images/kubernetes/container-files/openshift-brand [hunk 12]
```diff
@@ -0,0 +1,3 @@
+NAME=Cockpit Kubernetes
+ID=cockpit-kubernetes
+PRETTY_NAME=Cockpit Kubernetes
```
Comment (stefwalter) [inline images/kubernetes/container-files/openshift-brand]: This is backwards.

File: test/github-task [hunk 13]
```diff
@@ -215,6 +215,10 @@ def main():
         if opts.verbose:
             cmd.append("--verbose")
 
+    elif prefix == "images":
```
Comment (stefwalter) [inline test/github-task]: I would suggest "image" as in "image/kubernetes"

**Comment by stefwalter** [comment]
Other than that, looks good. Lets see how the tests do.

File: test/github-task [hunk 14]
```diff
@@ -215,6 +215,10 @@ def main():
...
+        cmd = [ "./images/run-tests", "--install", "--platform", value]
```
Comment (stefwalter) [inline test/github-task]: ./images/run-tests doesn't exist.

**Comment by stefwalter** [comment]
Checkout the exact HEAD commit here ... and then:

```
test/github-task --head --publish=fedorapeople.org image/kubernetes
```

Which produces:

```
First, rewinding head to replay your work on top of it...
Applying: images: cockpit container for kubernetes
Applying: test: Add ssl arguments to phantomjs
Applying: test: Add port argument to phantomjs browser
Applying: test: Make kube tests reusable
Applying: test: Add testing for containers
usage: run-tests [-h] [-j JOBS] [-v] [-t] [-q] [--thorough] [-s] [-i]
                 [-c CONTAINER]
                 [tests [tests ...]]
run-tests: error: unrecognized arguments: --platform
```

**Comment by stefwalter** [comment]
```
Building and installing Cockpit ...
usage: vm-download [-h] [--force] [--store STORE] [--prune] [-d]
                   [image [image ...]]
vm-download: error: image link does not exist: kubernetes
Traceback (most recent call last):
  File "./containers/run-tests", line 102, in <module>
    sys.exit(main())
  File "./containers/run-tests", line 90, in main
    install(opts)
  File "./containers/run-tests", line 61, in install
    subprocess.check_call(cmd)
  File "/usr/lib64/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/data/src/cockpit/test/vm-download', 'kubernetes']' returned non-zero exit status 2
```

**Pull request closed without merging**